src/midje/parsing/0_to_fact_form/generative.clj

```diff
@@ -71,19 +73,86 @@
           num-tests               (or (:num-tests opts) 10)
           quick-check-opts        (->> (select-keys opts [:seed :max-size])
                                        (into [])
-                                       flatten)]
-      `(let [fact-fn#       (fn ~prop-names
+                                       flatten)
+          fact-fn-name            (gensym "fact-fn")]
+      `(let [~fact-fn-name       (fn ~prop-names
                               ~(parse-facts/wrap-fact-around-body
                                  metadata checks))
-             prop#          (prop/for-all* ~prop-values fact-fn#)
+             prop#          (prop/for-all* ~prop-values ~fact-fn-name)
              [run# passes#] (run-for-all (list ~num-tests
                                                prop#
                                                ~@quick-check-opts))]
+         (if (:result run#)
+           (dotimes [_# (/ passes# ~num-tests)]
+             (emission/pass))
+           (run-with-smallest ~fact-fn-name '~prop-names run#))
+         (boolean (:result run#))))))
+
+(declare roll-up-bindings)
+
+(defn bind-expr [gen-name gen-expr inner-body other-bindings]
+  (roll-up-bindings
+    other-bindings
+    (list 'clojure.test.check.generators/bind
+          gen-expr
+          (list `fn [gen-name] inner-body))))
+
+(defn let-expr [fmap-bindings inner-body other-bindings]
+  (roll-up-bindings
+    other-bindings
+    `(let ~fmap-bindings
+       ~inner-body)))
+
+(defn roll-up-bindings [bindings inner-body]
+  (let [[fst-binding & rst-bindings] bindings
+        [lhs rhs] fst-binding]
+    (cond
+       (empty? bindings) inner-body
+       (= :let lhs)  (let-expr rhs inner-body rst-bindings)
+       :else (bind-expr lhs rhs inner-body rst-bindings))))
+
+
+(defn roll-up [bindings gen-names check-fn-name]
+  (roll-up-bindings (reverse bindings)
+                    `(gen/return {:function ~check-fn-name
+                                  :result   (~check-fn-name ~@gen-names)
+                                  :args     (list ~@gen-names)})))
+
+(defn variables [bindings]
```

Comment:
unit test or docstring to help with readability would be great. 
Also, this should be private along with the other new functions above (unless that messes with availability after macro expansion, which I don't think it should)